xcm/xcm-executor/src/lib.rs

```diff
@@ -17,116 +17,205 @@
 #![cfg_attr(not(feature = "std"), no_std)]
 
 use frame_support::{
-	dispatch::{Dispatchable, GetDispatchInfo},
+	dispatch::GetDispatchInfo,
 	ensure,
+	traits::{Contains, ContainsPair, Get, PalletsInfoAccess},
 };
+use parity_scale_codec::{Decode, Encode};
+use sp_io::hashing::blake2_128;
 use sp_runtime::traits::Saturating;
 use sp_std::{marker::PhantomData, prelude::*};
-use xcm::latest::{
-	Error as XcmError, ExecuteXcm,
-	Instruction::{self, *},
-	MultiAssets, MultiLocation, Outcome, Response, SendXcm, Weight, Xcm,
-};
+use xcm::latest::prelude::*;
 
 pub mod traits;
 use traits::{
-	ClaimAssets, ConvertOrigin, DropAssets, FilterAssetLocation, InvertLocation, OnResponse,
-	ShouldExecute, TransactAsset, VersionChangeNotifier, WeightBounds, WeightTrader,
+	validate_export, AssetExchange, AssetLock, CallDispatcher, ClaimAssets, ConvertOrigin,
+	DropAssets, Enact, ExportXcm, FeeManager, FeeReason, OnResponse, ShouldExecute, TransactAsset,
+	VersionChangeNotifier, WeightBounds, WeightTrader,
 };
 
 mod assets;
 pub use assets::Assets;
 mod config;
 pub use config::Config;
 
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct FeesMode {
+	pub jit_withdraw: bool,
+}
+
 /// The XCM executor.
 pub struct XcmExecutor<Config: config::Config> {
-	pub holding: Assets,
-	pub origin: Option<MultiLocation>,
-	pub original_origin: MultiLocation,
-	pub trader: Config::Trader,
+	holding: Assets,
+	holding_limit: usize,
+	context: XcmContext,
+	original_origin: MultiLocation,
+	trader: Config::Trader,
 	/// The most recent error result and instruction index into the fragment in which it occurred,
 	/// if any.
-	pub error: Option<(u32, XcmError)>,
+	error: Option<(u32, XcmError)>,
 	/// The surplus weight, defined as the amount by which `max_weight` is
 	/// an over-estimate of the actual weight consumed. We do it this way to avoid needing the
 	/// execution engine to keep track of all instructions' weights (it only needs to care about
 	/// the weight of dynamically determined instructions such as `Transact`).
-	pub total_surplus: u64,
-	pub total_refunded: u64,
-	pub error_handler: Xcm<Config::RuntimeCall>,
-	pub error_handler_weight: u64,
-	pub appendix: Xcm<Config::RuntimeCall>,
-	pub appendix_weight: u64,
+	total_surplus: u64,
+	total_refunded: u64,
+	error_handler: Xcm<Config::RuntimeCall>,
+	error_handler_weight: u64,
+	appendix: Xcm<Config::RuntimeCall>,
+	appendix_weight: u64,
+	transact_status: MaybeErrorCode,
+	fees_mode: FeesMode,
+	topic: Option<[u8; 32]>,
```

Comment:
`topic` is defined in two places: here and as part of `XvmContext` field.

`SetTopic` and `ClearTopic` operate on XCVM's `.topic` field but all the other commands just pass on reference to `context` field.

Should this one be removed? And aforementioned command handling updated to operate on `.context`?